flux-attrs/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+        )),
+        Ok(trait_item_method) => {
+            Ok((trait_item_method.sig, trait_item_method.attrs))
+        }
+    }
+}
+
+/// Creates a dummy function whose inner body is just a pass-through call to the
+/// extern function. Takes the signature and attributes
+/// ```ignore
+/// swap(a: &mut i32, b: &mut i32) -> ()
+/// ```
+/// and produces a function that looks like.
+/// ```ignore
+///     pub fn flux_extern_spec_swap(a: &mut i32, b: &mut i32) -> () {
+///         swap(a,b)
+///     }
+/// ```
+///
+/// The big idea:
+/// 1. Mangle the name of the function into flux_extern_spec_{fn_ident} so that it doesn't
+///    conflict with an existing function. TODO: add some random number to ensure it.
+/// 2. Create a new function with the mangled name and same signature.
+/// 3. Replace its body with a single expression that is a function call.
+/// 4. The function call is of the form mod_path::fn_ident<generic_args>(params)
+///    where the generic args and params have been modified so that they function as
+///    arguments to the call instead of being a part of the function signature.
+/// 5. Finally, we suppress dead_code/unused warnings and add the flux::extern_spec
+///    so that flux knows this is the generated dummy function for the external function
+///    that is being called.
+fn create_dummy_fn(
+    mod_path: Option<syn::Path>,
+    sig: syn::Signature,
+    attrs: Vec<syn::Attribute>,
+) -> syn::Result<TokenStream> {
+    let span = sig.span();
+    let mut mangled_sig = sig.clone();
+    let ident = sig.ident;
+    mangled_sig.ident = format_ident!("flux_extern_spec_{}", ident);
+    let generic_call_args = transform_generic_params_to_call_args(sig.generics.params);
+    let call_args = transform_params_to_call_args(sig.inputs);
+    let fn_path: syn::Path = if let Some(mod_path) = mod_path {
+        parse_quote_spanned!( ident.span() => #mod_path :: #ident)
+    } else {
+        parse_quote_spanned!( ident.span() => #ident )
+    };
+    Ok(quote_spanned! { span =>
+                        #[flux::extern_spec]
+                        #(#attrs)*
+                        #[allow(unused, dead_code)]
+                        #mangled_sig {
+                            #fn_path :: < #generic_call_args > ( #call_args )
+                        }
+    })
+}
+// Cribbed from Prusti's extern_spec_rewriter
+fn transform_generic_params_to_call_args(
+    generic_params: Punctuated<GenericParam, Token!(,)>,
+) -> Punctuated<GenericArgument, Token!(,)> {
+    Punctuated::from_iter(
+        generic_params
+            .iter()
+            .flat_map(|param| -> Option<GenericArgument> {
+                let span = param.span();
+                match param {
+                    GenericParam::Type(syn::TypeParam { ident, .. }) => {
+                        Some(parse_quote_spanned! { span => #ident })
+                    }
+                    GenericParam::Lifetime(_) => None,
```

Comment:
Why are we filtering out lifetimes here?

Reply:
We're transforming the function signature's arguments into arguments to pass to the external function. Do we need to explicitly indicate the lifetimes when they're call arguments?

Reply:
ah makes sense 👍🏽